File: 641. Design Circular Deque/solution.py
class Node:
    def __init__(self, val: int):
        self.val = val
        self.next = None
        self.prev = None

class MyCircularDeque:

    def __init__(self, k: int):
        self.totalSize = k
        self.currSize = 0
        self.front = None
        self.back = None

    def insertFront(self, value: int) -> bool:
        if self.currSize == self.totalSize:
            return False

        new_node = Node(value)

        if self.currSize == 0:
            self.front = self.back = new_node
        else:
            new_node.next = self.front
            self.front.prev = new_node
            self.front = new_node

        self.currSize += 1
        return True

    def insertLast(self, value: int) -> bool:
        if self.currSize == self.totalSize:
            return False

        new_node = Node(value)

        if self.currSize == 0:
            self.front = self.back = new_node
        else:
            self.back.next = new_node
            new_node.prev = self.back
            self.back = new_node

        self.currSize += 1
        return True

    def deleteFront(self) -> bool:
        if self.currSize == 0:
            return False

        tmp = self.front
        if self.front == self.back:
            self.front = self.back = None
        else:
            self.front = self.front.next
            self.front.prev = None

        self.currSize -= 1
        return True

    def deleteLast(self) -> bool:
        if self.currSize == 0:
            return False

        tmp = self.back
        if self.front == self.back:
            self.front = self.back = None
        else:
            self.back = self.back.prev
            self.back.next = None

        self.currSize -= 1
        return True

    def getFront(self) -> int:
        if self.currSize == 0:
            return -1
        return self.front.val

    def getRear(self) -> int:
        if self.currSize == 0:
            return -1
        return self.back.val

    def isEmpty(self) -> bool:
        return self.currSize == 0

    def isFull(self) -> bool:
        return self.currSize == self.totalSize

File: 641. Design Circular Deque/solution.js
class Node {
    constructor(val) {
        this.val = val;
        this.next = null;
        this.prev = null;
    }
}

/**
 * @param {number} k
 */
var MyCircularDeque = function(k) {
    this.totalSize = k;
    this.currSize = 0;
    this.front = null;
    this.back = null;
};

/** 
 * @param {number} value
 * @return {boolean}
 */
MyCircularDeque.prototype.insertFront = function(value) {
    if (this.currSize === this.totalSize) {
        return false;
    }

    const newNode = new Node(value);

    if (this.currSize === 0) {
        this.front = this.back = newNode;
    } else {
        newNode.next = this.front;
        this.front.prev = newNode;
        this.front = newNode;
    }

    this.currSize++;
    return true;
};

/** 
 * @param {number} value
 * @return {boolean}
 */
MyCircularDeque.prototype.insertLast = function(value) {
    if (this.currSize === this.totalSize) {
        return false;
    }

    const newNode = new Node(value);

    if (this.currSize === 0) {
        this.front = this.back = newNode;
    } else {
        this.back.next = newNode;
        newNode.prev = this.back;
        this.back = newNode;
    }

    this.currSize++;
    return true;
};

/**
 * @return {boolean}
 */
MyCircularDeque.prototype.deleteFront = function() {
    if (this.currSize === 0) {
        return false;
    }

    const tmp = this.front;
    if (this.front === this.back) {
        this.front = this.back = null;
    } else {
        this.front = this.front.next;
        this.front.prev = null;
    }

    this.currSize--;
    return true;
};

/**
 * @return {boolean}
 */
MyCircularDeque.prototype.deleteLast = function() {
    if (this.currSize === 0) {
        return false;
    }

    const tmp = this.back;
    if (this.front === this.back) {
        this.front = this.back = null;
    } else {
        this.back = this.back.prev;
        this.back.next = null;
    }

    this.currSize--;
    return true;
};

/**
 * @return {number}
 */
MyCircularDeque.prototype.getFront = function() {
    return this.currSize === 0 ? -1 : this.front.val;
};

/**
 * @return {number}
 */
MyCircularDeque.prototype.getRear = function() {
    return this.currSize === 0 ? -1 : this.back.val;
};

/**
 * @return {boolean}
 */
MyCircularDeque.prototype.isEmpty = function() {
    return this.currSize === 0;
};

/**
 * @return {boolean}
 */
MyCircularDeque.prototype.isFull = function() {
    return this.currSize === this.totalSize;
};